rust_src/src/crypto/mod.rs

```diff
@@ -1,78 +1,470 @@
+            let coding_system = validate_coding_system(
+                get_coding_system_for_buffer(
+                    object,
+                    buffer.unwrap(),
+                    start,
+                    end,
+                    start_byte,
+                    end_byte,
+                    coding_system,
+                ),
+                noerror,
+            );
+            LispObject::from_raw(unsafe {
+                code_convert_string(
+                    s.to_raw(),
+                    coding_system.to_raw(),
+                    LispObject::constant_nil().to_raw(),
+                    true,
+                    false,
+                    false,
+                )
+            }).as_string_or_error()
+        } else {
+            s.as_string_or_error()
+        }
+    } else {
+        unsafe {
+            wrong_type_argument(Qstringp, object.to_raw());
+        }
+    }
+}
+
+/// Return the secure hash of OBJECT, a buffer or string.
+/// ALGORITHM is a symbol specifying the hash to use:
+/// md5, sha1, sha224, sha256, sha384 or sha512.
+///
+/// The two optional arguments START and END are positions specifying for
+/// which part of OBJECT to compute the hash.  If nil or omitted, uses the
+/// whole OBJECT.
+///
+/// If BINARY is non-nil, returns a string in binary form.
+#[lisp_fn(min = "1")]
+fn md5(
+    object: LispObject,
+    start: LispObject,
+    end: LispObject,
+    coding_system: LispObject,
+    noerror: LispObject,
+) -> LispObject {
+    let mut string = object.as_string();
+    let buffer = object.as_buffer();
+    let input = get_input(
+        object,
+        &mut string,
+        &buffer,
+        start,
+        end,
+        coding_system,
+        noerror,
+    );
+    _secure_hash(HashAlg::HashMD5, input.as_slice(), true)
+}
+
+/// Return the secure hash of OBJECT, a buffer or string.
+/// ALGORITHM is a symbol specifying the hash to use:
+/// md5, sha1, sha224, sha256, sha384 or sha512.
+///
+/// The two optional arguments START and END are positions specifying for
+/// which part of OBJECT to compute the hash.  If nil or omitted, uses the
+/// whole OBJECT.
+///
+/// If BINARY is non-nil, returns a string in binary form.
+#[lisp_fn(min = "2")]
+fn secure_hash(
+    algorithm: LispObject,
+    object: LispObject,
+    start: LispObject,
+    end: LispObject,
+    binary: LispObject,
+) -> LispObject {
+    let mut string = object.as_string();
+    let buffer = object.as_buffer();
+    let input = get_input(
+        object,
+        &mut string,
+        &buffer,
+        start,
+        end,
+        LispObject::constant_nil(),
+        LispObject::constant_nil(),
+    );
+    _secure_hash(hash_alg(algorithm), input.as_slice(), binary.is_nil())
+}
 
 #[no_mangle]
-pub unsafe extern "C" fn sha1_buffer(
-    buffer: *const c_char,
-    len: size_t,
-    dest_buf: *mut c_void,
-) -> *mut c_void {
+fn _secure_hash(algorithm: HashAlg, input: &[u8], hex: bool) -> LispObject {
```

Comment:
If it's `no_mangle`, it should also be `extern "C"`. But I guess the function is not called from C anymore?